### ISSUE TITLE
[Android] CollectionView Crashes when ItemSource Changed 

### ISSUE THREAD

**Issue by minaairsupport** [xamarin/Xamarin.Forms#5110]
### Description
App Crash when ItemSource Changed for CollectionView 

### Steps to Reproduce

1. Create ContentPage contain CollectionView 
2. init item source for CollectionView with an empty list 
3. Load data from web service and make any delay then fill this ItemSource 

### Expected Behavior
ItemSource Changed 

### Actual Behavior
App Crashes with this error 
```
JniEnvironment+InstanceMethods.CallNonvirtualVoidMethod (Java.Interop.JniObjectReference instance, Java.Interop.JniObjectReference type, Java.Interop.JniMethodInfo method, Java.Interop.JniArgumentValue* args)
Java.Lang.IllegalStateException: Observer md51558244f76c53b6aeda52c8a337f2c37.DataChangeObserver@e23f401 was not registered.
```

### Basic Information

- Version with issue: 4.0.0.94569-pre3
- Last known good version:
- IDE: Visual Studio Mac 2017 , 2019 
- Platform Target Frameworks: 
  - Android: 7.0
- Nuget Packages:4.0.0.94569-pre3
- Affected Devices:Nexsus 4 , Android 7.0 api 24

**Comment by kingces95** [comment]
Can you please attach a reproduction? I'm guessing how and where the delay is injected will affect whether this reproduces.

**Comment by rubenglam** [comment]
Same error 👍 . I have a searchView that filters the items and when there are 0 items, just after modify the query to that it contains items the app crash with the same error.

**Comment by davidbritch** [comment]
@kingces95 I've been able to repro this in 4.0-pre5. Setting the `ItemsSource` property after a 2 second delay causes a crash:

```
        protected override async void OnAppearing()
        {
            base.OnAppearing();

            await Task.Delay(2000);
            collectionView.ItemsSource = (BindingContext as MonkeysViewModel).Monkeys;
        }
```

I tried all kinds of threading options, and marshalling to the UI thread (when appropriate) and it always happens.

## Repro steps

1. Run the attached sample (either iOS or Android).
2. Browse to the `EmptyViewLoadSimulationPage`.
3. Wait 2 seconds.

## Actual result

Sample crashes.

## Expected result.

Sample updates the `CollectionView` from an `ObservableCollection`.

## Repro link

[CollectionViewDemos.zip(https://github.com/xamarin/Xamarin.Forms/files/2962251/CollectionViewDemos.zip)

## iOS Crash

`MonoTouchException`:

![Screen Shot 2019-03-13 at 14 31 33](https://user-images.githubusercontent.com/8092460/54292821-bd0acc80-45a6-11e9-989e-6899ed654f4c.png)

## Android crash

`IllegalStateException`:

![Screen Shot 2019-03-13 at 14 34 29](https://user-images.githubusercontent.com/8092460/54292864-d4e25080-45a6-11e9-9991-6f9c2bb9a2ca.png)

**Comment by davidbritch** [comment]
@hartez I've retried this on 3.6 and 4.0-pre6 and the results are identical for both. On iOS a `MonoTouchException` occurs, as per the screenshot above. On Android it works fine.

**Comment by hartez** [comment]
This crash is fixed for Android in 4.0.0-pre6 and for iOS in 4.0.0-pre8.
